app/controllers/decidim/extended_socio_demographic_authorization_handler/autocomplete_controller.rb

```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+module Decidim
+  module ExtendedSocioDemographicAuthorizationHandler
+    class AutocompleteController < ApplicationController
+      def city
+        return render json: [] if params[:postal_code].blank?
+
+        return render json: Rails.cache.read(cache_key) if Rails.cache.exist?(cache_key)
+
+        response = Decidim::ExtendedSocioDemographicAuthorizationHandler::AutocompleteService.for(params[:postal_code])
+
+        return render json: [] if response == "null"
+
+        Rails.cache.write(cache_key, response, expires_in: 1.month)
+
+        render json: response
+      end
+
+      def cache_key
+        "postal_code_autocomplete/#{params[:postal_code]}"
+      end
```

Comment:
IMHO a controller should juste fetch and render, here it is managing the cache state, expiration and logic about the API response.

You should abstract it into the service like in this example:

https://github.com/OpenSourcePolitics/decidim-module-extended_socio_demographic_authorization_handler/blob/hotfix/disable_autocomplete/app/services/opendatasoft_service.rb